### ISSUE TITLE
Aerosol-related PET log error message

### ISSUE THREAD

**Issue by DeniseWorthen** [ufs-community/ufs-weather-model#1888]
## Description

I compiled and ran both the ``cpld_control_nowave_noaero_p8`` and ``cpld_bmark_p8`` in debug mode using @ulmononian's feature/add_c5 branch. Both tests run, but the bmark test (which has aerosols) produces an PET log error message on PEs 000 through 0766. There is no error message in the 0767 log. The aerosol component runs on PEs 0:767. I don't know why the last Aerosol PE does not produce the error.

The error message is:

```
20230907 124949.789 INFO             PET0766 UFS Aerosols: Advancing from 2013-04-01T00:00:00 to 2013-04-01T00:05:00
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:668 Info::erase() Not found  - [json.exception.out_of_range.403] key 'GridCornerLons:' not found
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:688 Info::erase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMC_InfoCDef.C:243 ESMC_InfoErase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMF_Info.F90:2656 ESMF_InfoRemove() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 src/Superstructure/AttributeAPI/interface/ESMF_Attribute.F90:46022 ESMF_AttributeRemoveAttPackGrid( Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:668 Info::erase() Not found  - [json.exception.out_of_range.403] key 'GridCornerLats:' not found
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:688 Info::erase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMC_InfoCDef.C:243 ESMC_InfoErase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMF_Info.F90:2656 ESMF_InfoRemove() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 src/Superstructure/AttributeAPI/interface/ESMF_Attribute.F90:46022 ESMF_AttributeRemoveAttPackGrid( Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:668 Info::erase() Not found  - [json.exception.out_of_range.403] key 'GridCornerLons:' not found
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:688 Info::erase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMC_InfoCDef.C:243 ESMC_InfoErase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMF_Info.F90:2656 ESMF_InfoRemove() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 src/Superstructure/AttributeAPI/interface/ESMF_Attribute.F90:46022 ESMF_AttributeRemoveAttPackGrid( Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:668 Info::erase() Not found  - [json.exception.out_of_range.403] key 'GridCornerLats:' not found
20230907 125011.127 ERROR            PET0766 ESMCI_Info.C:688 Info::erase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMC_InfoCDef.C:243 ESMC_InfoErase() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 ESMF_Info.F90:2656 ESMF_InfoRemove() Not found  - Internal subroutine call returned Error
20230907 125011.127 ERROR            PET0766 src/Superstructure/AttributeAPI/interface/ESMF_Attribute.F90:46022 ESMF_AttributeRemoveAttPackGrid( Not found  - Internal subroutine call returned Error
20230907 125021.303 INFO             PET0766 Model Advance: before wrtcomp run
```

### To Reproduce:

compile and run the feature/add_c5 branch using DEBUG for the bmark P8 test.

**Comment by ulmononian** [comment]
i had 100% of tests pass about a week ago or so on c5. so the debug test runs to completion here but you are seeing these errors anyway?

**Comment by DeniseWorthen** [comment]
Yes, the test runs, so it is not fatal. I don't know what it means actually. It must be in MAPL?

**Comment by natalie-perlin** [comment]
@DeniseWorthen @ulmononian - 
does spack-stack need to have esmf-debug module for running the  cpld_control_nowave_noaero_p8 and cpld_bmark_p8 in debug mode ?

**Comment by DeniseWorthen** [comment]
No, we no longer use esmf built w/ debug.

**Comment by zach1221** [comment]
Hi, @DeniseWorthen . Below are my experiment directories for cpld_control_nowave_noaero_p8_intel & cpld_bmark_p8_intel on Gaea C5, if you want to take a look at the PET logs. If you dont have access to view let me know.

/lustre/f2/scratch/Zachary.Shrader/FV3_RT/rt_76751/cpld_control_nowave_noaero_p8_intel
/lustre/f2/scratch/Zachary.Shrader/FV3_RT/rt_76751/cpld_bmark_p8_intel

**Comment by DeniseWorthen** [comment]
@zach1221 The same behaviour appears to be present. No PET error messages in the ``cpld_control_nowave_noaero_p8_intel`` test, which does not include aerosols. In the ``cpld_bmark_p8_intel`` test, the only PET log which does **not** contain the error message is PET0767 (the last atm PET).

**Comment by natalie-perlin** [comment]
@DeniseWorthen @zach1221 
- all RT tests passes for Gaea-c5 with spack-stack/1.5.0.
 Modulefiles and scripts for gaea-c5 were updated as continuation of the PR into the UFS weather model repo started by @ulmononian (ufs-community/ufs-weather-model#1784).
 A PR with the updates was created: https://github.com/ulmononian/ufs-weather-model/pull/13
into @ulmononian fork, the log [RegressionTests_gaea-c5.log](https://github.com/ulmononian/ufs-weather-model/files/13364679/RegressionTests_gaea-c5.log) is attached to that PR.

**Comment by DeniseWorthen** [comment]
@zach1221 @jkbk2004 The behavior on hercules w/rt these two cases is the same on Hercules. That is, there is no ERROR on the test w/o aerosols, ie, cpld_control_noaero_p8_intel. There is an PET log error on all but the last ATM PE for the cpld_bmark_p8 test. I also checked the low-resolution case, cpld_control_p8_intel and it shows the same thing---ERROR on all but the last ATM PE. Note, Aerosols run on the same PEs as the ATM. I believe this most likely a MAPL issue and not ESMF.

**Comment by natalie-perlin** [comment]
@DeniseWorthen - 
   A working directory on Gaea-c5 with the cpld_bmark_p8 test is /lustre/f2/scratch/ncep/Natalie.Perlin/FV3_RT/rt_221558/cpld_bmark_p8_intel
  In case you may take a look to see whether this is the same behavior.

**Comment by DeniseWorthen** [comment]
@natalie-perlin I had checked earlier w/ @zach1221 run directories and confirmed that the same error message is present as in the original issue. I also later confirmed that the same message is present on a hercules run so it isn't specific to C5. I believe it to be a MAPL issue most likely.

**Comment by jkbk2004** [comment]
I think GSFC people might be able to access to hercules. It will be a great starting point if we can set up a MAPL debugging installation and experiment. @mathomp4 Is Jiang still available?

**Comment by mathomp4** [comment]
> I think GSFC people might be able to access to hercules. It will be a great starting point if we can set up a MAPL debugging installation and experiment. @mathomp4 Is Jiang still available?

@jkbk2004 Yes. @weiyuan-jiang should be able to help. That said, I'm not sure he (or any of us) have access to Hercules. We do have access to Orion.

**Comment by weiyuan-jiang** [comment]
I think I have access to Hercules. Judging from the logging error, it is more like a problem from  ESMF ( due to the building ? ). But I will take a look at it.

**Comment by DeniseWorthen** [comment]
@weiyuan-jiang Thanks for checking. The reason I suspect MAPL is that we do not see the error in cases w/o the aerosol component.

**Comment by weiyuan-jiang** [comment]
What version of gocart and MAPL are you using? @DeniseWorthen

**Comment by DeniseWorthen** [comment]
I'm not sure how to tell which version of gocart we are using, but for MAPL we are using 2.35.2-esmf-8.4.2

**Comment by jkbk2004** [comment]
GOCART hash: https://github.com/GEOS-ESM/GOCART/tree/041422934cae1570f2f0e67239d5d89f11c6e1b7

**Comment by zach1221** [comment]
Looks like 2.1.1 for gocart

**Comment by weiyuan-jiang** [comment]
I have talked to Ben. He thought there might be a small chance that an attribute is not there in the MAPL_GetHorzIJIndex call. ( It is confusing though, because it should not produce error) . Anyway, I have replaced that call in the [branch](https://github.com/GEOS-ESM/MAPL/tree/feature/wjiang/replace_GetHorzIJindex)  . Would you please try this new MAPL?

**Comment by DeniseWorthen** [comment]
Since this error is not specific to C5 (it also shows up on Hercules), I've edited the issue title.

**Comment by NickSzapiro-NOAA** [comment]
@DeniseWorthen I see this error in GEFS RT intel debug mode

**Comment by DeniseWorthen** [comment]
Just verifying that this is still an issue. Perhaps some random MAPL update will finally resolve it.

**Comment by weiyuan-jiang** [comment]
@DeniseWorthen Would you please tell us which version or tag MAPL you are using?

**Comment by DeniseWorthen** [comment]
@weiyuan-jiang In UWM, we're still using SS16, so we have ``"[mapl]"           = "2.40.3-esmf-8.6.0"}`` in ufs_common

**Comment by mathomp4** [comment]
@DeniseWorthen We have a thought. Can you try the branch `bugfix/mapl-2.40.6` of MAPL?

You can see the change here:

https://github.com/GEOS-ESM/MAPL/compare/v2.40.5...bugfix/mapl-2.40.6?w=1

We aren't sure, but your top most error seems to say it's removing an Attribute that doesn't exist, so we are just testing that first.

**Comment by DeniseWorthen** [comment]
@mathomp4 I'm not set up or experienced enough in SS building to test different MAPL versions. I know our next update should be @RatkoVasic-NOAA 's PR for SS191 which (it appears) to use 2.53.3. Would any fix for this be included in that?

**Comment by mathomp4** [comment]
No. This isn't in our current line of development (the code changed is pretty much the same as it is now)...but we can add it. I'll need to consult people here but if they are happy, I can port it to 2.53.

**Comment by mathomp4** [comment]
@DeniseWorthen Okay. I can make a MAPL 2.53.4 but for @RatkoVasic-NOAA would you be able to add it to this system? (Or to SS16?)

I can make a patched tag I'd just need to know where.

**Comment by RatkoVasic-NOAA** [comment]
> [@DeniseWorthen](https://github.com/DeniseWorthen) Okay. I can make a MAPL 2.53.4 but for [@RatkoVasic-NOAA](https://github.com/RatkoVasic-NOAA) would you be able to add it to this system? (Or to SS16?)
> 
> I can make a patched tag I'd just need to know where.

Yes, I can install 2.53.4 on Tier 1 platforms as soon as it is ready. And update my UFS-WM branch (ss-191) to use it.

**Comment by mathomp4** [comment]
@RatkoVasic-NOAA I present to you [MAPL 2.53.4](https://github.com/GEOS-ESM/MAPL/releases/tag/v2.53.4).

I have a PR into mainline spack as well: https://github.com/spack/spack/pull/50515

**Comment by mathomp4** [comment]
@DeniseWorthen Any news yet on if this helped? We have our fingers crossed 🤞🏼

**Comment by DeniseWorthen** [comment]
@mathomp4 Oops, I didn't realize you were waiting for me to test. Let me check the SS191 PR and make sure it has been put in and then I can test.

**Comment by mathomp4** [comment]
@DeniseWorthen Oh. No big rush, just wanted to make sure the fix...fixed. 🙂

**Comment by RatkoVasic-NOAA** [comment]
@DeniseWorthen @mathomp4 my plan was to put it in 1.9.1 today. Will let you know when it is done.

**Comment by RatkoVasic-NOAA** [comment]
@DeniseWorthen and @mathomp4 
MAPL@2.53.4 is installed on: Hera, Gaea-C6, Derecho, Hercules and Orion
modulefile name is: 2.53.4-esmf-8.8.0.lua

**Comment by DeniseWorthen** [comment]
@RatkoVasic-NOAA I merged in your SS191 branch on Hera, changed the ufs_common to point to the new mapl/esmf and tried to compile. I got a whole bunch of errors. I didn't paste them all---it seems like the same ones were reported multiple times (?).

```
 ./compile.sh hera "-DAPP=S2SWA -DDEBUG=ON -DCCPP_SUITES=FV3_GFS_v17_coupled_p8_ugwpv1" test.intel.db intel NO NO 2>&1 | tee test.intel.db.log

CMake Error at /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/mapl-2.53.4-dwuvax4/lib64/cmake/MAPL/MAPL-targets.cmake:70 (set_target_properties):
  The link interface of target "MAPL_cfio_r4" contains:

    ESMF::ESMF

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/mapl-2.53.4-dwuvax4/lib64/cmake/MAPL/mapl-config.cmake:74 (include)
  GOCART/CMakeLists.txt:70 (find_package)


CMake Error at /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/mapl-2.53.4-dwuvax4/lib64/cmake/MAPL/MAPL-targets.cmake:122 (set_target_properties):
  The link interface of target "MAPL.generic" contains:

    ESMF::ESMF

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /contrib/spack-stack/spack-stack-1.9.1/envs/ue-oneapi-2024.2.1/install/oneapi/2024.2.1/mapl-2.53.4-dwuvax4/lib64/cmake/MAPL/mapl-config.cmake:74 (include)
  GOCART/CMakeLists.txt:70 (find_package)

....

```

**Comment by RatkoVasic-NOAA** [comment]
@DeniseWorthen although I don't think it will help in compilation stage, there is one manual change in GOCART that has to be applied to my branch:
Hera: /scratch2/NCEPDEV/fv3-cam/Ratko.Vasic/WM-SS-191-RV/GOCART/Process_Library/GOCART2G_Process.F90

Also, does it work if you just return mapl to 2.53.3?

**Comment by DeniseWorthen** [comment]
@RatkoVasic-NOAA I forgot that your branch had updated GOCART; I updated that (+ your manual fix) and it is compiling now.

**Comment by DeniseWorthen** [comment]
Good news, the PET log messages are gone. Bad news, the code still seg-faults. This is w/ your manual fix you pointed me to.

```
1: [h25c08:1242693:0:1242693] Caught signal 8 (Floating point exception: floating-point divide by zero)
143: ==== backtrace (tid: 832181) ====
143:  0 0x0000000000053519 ucs_debug_print_backtrace()  ???:0
143:  1 0x0000000000012990 __funlockfile()  :0
143:  2 0x0000000002a662d4 gocart2g_process_mp_settlingsolver_()  /scratch1/NCEPDEV/nems/Denise.Worthen/WORK/ufs_dw/GOCART/Process_Library/GOCART2G_Process.F90:1927
143:  3 0x0000000002a4be97 gocart2g_process_mp_chem_settlingsimple_()  /scratch1/NCEPDEV/nems/Denise.Worthen/WORK/ufs_dw/GOCART/Process_Library/GOCART2G_Process.F90:1580
```

I tried to compile/run GNU but the compile failed on the cmake, jasper and nccmp versions.

**Comment by weiyuan-jiang** [comment]
@DeniseWorthen You probably need this fix  https://github.com/GEOS-ESM/GOCART/pull/323

**Comment by DeniseWorthen** [comment]
@weiyuan-jiang Currently I'm using Ratko's manual fix; I then updated manually for the 323 PR you reference above. 

I noticed there are a couple of additional line changes and I'm getting confused about what all the modifications are. This my git diff between the hash I'm pointing to 

```
...
[submodule "GOCART"]
  path = GOCART
  url = https://github.com/GEOS-ESM/GOCART
  branch = develop
....

 837737f (HEAD, origin/develop) - (6 weeks ago) Merge pull request #314 from GEOS-ESM/feature/wjiang/fix_#313 - Virginie Buchard
```
and my local mods

```
diff --git a/Process_Library/GOCART2G_Process.F90 b/Process_Library/GOCART2G_Process.F90
index 35ef722..db9708c 100644
--- a/Process_Library/GOCART2G_Process.F90
+++ b/Process_Library/GOCART2G_Process.F90
@@ -1916,11 +1916,9 @@ end function DarmenovaDragPartition
           dp_  = delp(i,j,:)
           tau_ = tau(i,j,:)

-          dt_cfl  = 1 / maxval(tau_)
+          dt_cfl = abs(1.0 / maxval(tau_))

-          if (dt_cfl <= 0.) then
-             nSubSteps = 0
-          else if (dt_cfl > cdt) then
+          if (dt_cfl > cdt) then
              ! no need for time sub-splitting
              nSubSteps = 1
              dt = cdt
@@ -2436,7 +2434,7 @@ end function DarmenovaDragPartition
           qmin = minval(dz(i,j,:)/vsettle(i,j,:))
           minTime = min(cdt,qmin)
       !   Now, how many iterations do we need to do?
-          if ( minTime <= 0 ) then
+          if ( minTime < 0 ) then
              nSubSteps = 0
           else if(minTime .ge. cdt) then
              nSubSteps = 1
@@ -2776,7 +2774,7 @@ end function DarmenovaDragPartition
     minTime = min(minTime,qmin)

 !   Now, how many iterations do we need to do?
-    if ( minTime <= 0 ) then
+    if ( minTime < 0 ) then
          nSubSteps = 0
 !         call mpout_log(myname,'no Settling because minTime = ', minTime )
```

Where are the changes to minTime coming from? Are they correct or not? 

My current GOCART local changes *do* pass the debug test, and all the log messages are gone. My build is here if you want to check the GOCART

```
/scratch1/NCEPDEV/nems/Denise.Worthen/WORK/ufs_dw
```

**Comment by weiyuan-jiang** [comment]
I happened to see the minTime and made a change to avoid a potential crash when minTime =0. It may never happen though. My change would not  hurt.

**Comment by DeniseWorthen** [comment]
So will the change to minTime be included in a PR to the upstream? It seems it must be coming from the manual fix that I grabbed from Ratko.
